### PR TITLE
New endpoints to generate and check analysis reports

### PR DESCRIPTION
**Pull request by AlastairTaft**
_Created a PR so that I can make comments_

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -109,6 +113,54 @@ const getSalt = async event => {
...
+/**
+ * Allows an authority to report an analysis
+ * @return {Promise<string>} Id of the report
+ */
+const reportAnalysis = async event => {
+  let authorization = event.headers['authorization']
+  if (!authorization) throw new ServerError("Missing header 'authorization'", 400)
+
+  let { date } = JSON.parse(event.body)
+  date = moment(date)
```
Comment (AlastairTaft) [inline server/index.js:125]: Is adding `moment` a bit heavy weight for validating a date?

What about `isNaN(new Date(date)))`?

File: server/db/analysisReports.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
+const AnalysisReport = require('./model/AnalysisReport')
+
+const randomNumbers = (size) => Math.random().toString().substring(2, size + 2)
+
+const generateUuid = async (db) => {
+  const uuid = `${randomNumbers(4)}-${randomNumbers(4)}-${randomNumbers(2)}`
```
Comment (AlastairTaft) [inline server/db/analysisReports.js:6]: Not sure this is random enough. It could probably be brute forced quite easily.

File: server/index.js [hunk 3]
```diff
@@ -109,6 +113,54 @@ const getSalt = async event => {
...
+
+  const { db, client } = await getConnection()
+
+  // TODO: Improve authorization mechanism. Current weak implementation is for hackathon demo purposes
+  const valid = await dbAuthorities.validToken(db, authorization)
+  if(!valid) throw new ServerError("The authority is not valid'", 400)
+
+  const reportId = await dbAnalysisReports.saveReport(db, date)
+    .catch(() => {
+      throw new ServerError("Something went wrong reporting the analysis", 500)
```
Comment (AlastairTaft) [inline server/index.js:136]: Perhaps `console.error` the real error so we're not eating errors here?

File: server/index.js [hunk 4]
```diff
@@ -109,6 +113,54 @@ const getSalt = async event => {
...
+ * Check the analysis report
+ * @return {Promise<string>} Analysis report
+ */
+const checkAnalysisReport = async event => {
+  const uuid = event.queryStringParameters && event.queryStringParameters['uuid']
+  if (!uuid) throw new ServerError("Missing 'uuid' prop.", 400)
+
+  const { db, client } = await getConnection()
+
+  const reportId = await dbAnalysisReports.getReport(db, uuid)
```
Comment (AlastairTaft) [inline server/index.js:154]: Perhaps rename to `report` as I don't think this is an id here?
```suggestion
  const report = await dbAnalysisReports.getReport(db, uuid)
```

File: server/index.js (hunk 3 above)
Comment (Serg4554) [reply server/index.js:136]: Sure!

File: server/index.js (hunk 1 above)
Comment (Serg4554) [reply server/index.js:125]: You are right, I added to manage dates easier, but in the end I only used it for that

File: server/index.js (hunk 4 above)
Comment (Serg4554) [reply server/index.js:154]: Actually is the UUID of the report inserted in the database (not the complete report info), as we only want the endpoint to return the UUID to generate the QR. I will rename it to `reportUuid` 👍

File: server/db/analysisReports.js (hunk 2 above)
Comment (Serg4554) [reply server/db/analysisReports.js:6]: Agree, will replace it for a CSPRNG generator

File: server/db/analysisReports.js [hunk 5]
```diff
@@ -1,21 +1,26 @@
...
-  const uuid = `${randomNumbers(4)}-${randomNumbers(4)}-${randomNumbers(2)}`
-  let unique
+  let uuid, unique
+
   do {
+    const high = (await randomNumber(0, 9999)).toString().padStart(4, '0')
+    const mid = (await randomNumber(0, 9999)).toString().padStart(4, '0')
+    const low = (await randomNumber(0, 9999)).toString().padStart(4, '0')
+
+    uuid = `${high}-${mid}-${low}`
```
Comment (AlastairTaft) [inline server/db/analysisReports.js:12]: Still quite easy to brute force (I think), what about generating strings like this

```
5nbxKPSZ-LR$-j0q8LpdJyoPX3iylv9TH9O0YZyN9VPpE6_nNTl3k4RkgNTJ7unQ9$1L7DU*BsFPMbgZ-78S
```
I wrote a little utility function that could do it.

```js
const generateKey = function(bits){
  var alphabet = "012345678910abcdefghijklmnopqrstuvwxyz" + 
    "ABCDEFGHIJKLMNOPQRSTUVWXYZ$-_+!*"
  var expectedEntroy = Math.pow(2, bits)
  var key = ''
  while(Math.pow(alphabet.length, key.length) < expectedEntroy){
    key += alphabet[Math.floor(Math.random() * alphabet.length)]
  }
  return key
}

// Will result in at least 512 bits of entropy 
console.log(generateKey(512))
```

**Pull request closed without merging**